### PR TITLE
[FEATURE] - Late Unregistrations Handling V2

### PR DESCRIPTION
**Pull request by o-rayer**
# Feature aiming at reducing the number of late unregistrations to events

## Functional Schema
Capture d’écran 2024-09-02 à 17 31 02
Capture d’écran 2024-09-02 à 17 31 10

- An Unregistration is considered late if the event has a required activity (ex : collectives) and if the user unregisters less than 48h (parameterized in `configuration.yaml`) prior to the Event starting time

## Key Features

- Leverages the Badges logic to assign 2 types of Badges : `LateUnregisterWarning` & `Banned`
- `LateUnregisterWarning` expires the next 30th of September
- `Banned` expires 4 weeks after assignment
- When a user has a valid `Banned` badge, he cannot register to an event
- 'Se Désinscrire' button now has a proper pop-up screen instead of a flash, personalized if is it whether a late unsubscription or not
- When unregistering lately, a warning or banishment email is sent to the user by email, according to its warning level
- An Admin can manage badges and can handle banishment if required, using badge or user admin panels
- A user can visualize his warning status (Aucun avertissement, 1/2 avertissement(s), BANNI) in "Mon profil"
Capture d’écran 2024-09-02 à 17 22 12
- The "level" for Unregistration-related badges is now simply an iterative number for the number of given badges over time
- Stats module has a new line chart following Unregistrations, LateUnregistrations & UnjustifiedAbsentees per week over the last 52 weeks
- Updated db models for later features : Registrations have a `registration_time` datetime field, Badges have a `creation_time` datetime field

## Things to review or determine in priority

- The whole functional logic
- Transition table possibilities for late unregistrations
- Multi-day events unregistration
- All mails & pop-up contents and design choices
- If mailing fails, the logic should still work

## Next ideas / Things to do

- **Leader actions** : transition by a from “LateSelfUnregistered” to “Active” should remove the latest Warning Badge assigned (by default the most valid & restrictive one)
- **Leader actions** : Unjustified Absentees should be handled in the Ban logic
- **Leader actions** : remove latest Warning Badge assigned (by default the most valid & restrictive one) to "registered" users if an event is cancelled last minute
- Unregistering to Multi-days events should be considered late 7 days before event start (and not 48h)
- A User should see somewhere the remaining banishment time (account suspension) or the number of warnings issues or remaining
- Mailing functional testing
- Warning mails must be sent to : responsable d'activité, au participant, aux encadrants des collectives sur lequel il est inscrit actuellement
- Rationalize tests using proper factories to generate users, events etc...
- Add a `last_unregistration_time` or `last_self_unregistration_time` to the `Registration` object in order to follow unregistration properly in the stats module
Capture d’écran 2024-02-13 à 12 11 54

## Known issues

- Unauthenticated users test is failing
- Unsure of db default values for new fields
- On the event page, if a user if late unregistered and banned, messages are mixed-up : 
*Les inscriptions en ligne sont possibles lundi 2 septembre 2024 au mardi 3 septembre 2024.*
*Vous êtes inscrit à cet évènement en tant que Participant avec le statut Auto désinscrit tardif.*
*Vous vous êtes désincrit tardivement. Si vous souhaitez vous réinscrire, contactez l'encadrant.*
*Vous êtes actuellement banni suite à un trop grand nombre de désinscriptions tardives. Vous pourrez vous ré-inscrire à nouveau 4 semaines après votre dernière désinscription tardive. Si vous pensez que c'est une erreur, merci de contacter un administrateur.*

# Update https://github.com/Club-Alpin-Annecy/collectives/pull/721/commits/76c5b680e6ee7091b57f628f64c12025e7406fb5

## Functional Schema
Capture d’écran 2024-10-03 à 15 22 13

## Key features
- `Badge` and `Registration` models are now linked
- When a leader assigns a `UnJustifiedAbsentee` status, the Warning logic is applied to the user
- When a user justifies his absence and a leader assigns a `JustifiedAbsentee` status from a `UnjustifiedAbsentee`, all warning badges related to the Registration are deleted
- When a leader assigns any registration status (except for `UnjustifiedAbsentee`) from a `LateSelfUnregistered` status, all warning badges related to the Registration are deleted
- If a leader deletes an event, all Warning badges related to the event (ie Registrations of users to this event) are deleted

### CODE REVIEW

**Comment by gdaviet** [comment]
@o-rayer j'essaie de faire une review du code dans la semaine, en tout ca la partie fonctionnelle me parait très bien -- et merci pour les schémas, on pourra les ajouter à la doc

File: collectives/models/event/registration.py [hunk 1]
```diff
@@ -216,6 +217,38 @@ def is_unregistered(self, user):
...
+        """Check if an event starts within xh of current time.
+
+        :return: True if event starts within xh of current time
+        :rtype: boolean
+        """
+        return self.start < datetime.now() + timedelta(
+            hours=Configuration.LATE_UNREGISTRATION_THRESHOLD
+        )
+
+    def is_banned(self, user):
```
[the remark this answers is not in the document]
Comment (o-rayer) [reply collectives/models/event/registration.py:242]: Yes, après je me demandais si ça pouvait break la logique de `event.html` du coup mais je vais essayer

File: collectives/models/user/badge.py [hunk 2]
```diff
@@ -91,3 +109,114 @@ def activities_with_valid_badge(
...
+        """
+        badge_data = {}
+        if expiration_date is not None:
+            badge_data["expiration_date"] = expiration_date
+        if activity_id is not None:
+            badge_data["activity_id"] = activity_id
+        if level is not None:
+            badge_data["level"] = level
+
+        Badge.query.filter_by(user_id=self.id, badge_id=badge_id).update(badge_data)
```
[the remark this answers is not in the document]
Comment (o-rayer) [reply collectives/models/user/badge.py:163]: Tu as raison, en fait je pense que c'est misleading, et je n'utilise plus ces fonctions donc je vais les enlever

File: collectives/models/user/badge.py [hunk 3]
```diff
@@ -91,3 +109,114 @@ def activities_with_valid_badge(
...
+                    date.today().year
+                    if date.today().month < 10
+                    else date.today().year + 1,
+                    9,
+                    30,
+                )
+                self.assign_badge(
+                    badge_id,
+                    expiration_date=expiration_date,
+                    level=num_warning_badges + 1,
```
[the remark this answers is not in the document]
Comment (o-rayer) [reply collectives/models/user/badge.py:216]: Le `level` n'est plus utilisé au niveau fonctionnel pour le moment, c'est simplement du comptage. Pour les retraits de badges je vais associer badge et registration pour avoir un lien fort, notamment pour les annulations d'event par ex.

File: migrations/versions/01e5792cdfac_add_registration_time.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table("badges", schema=None) as batch_op:
+        batch_op.alter_column(
+            "badge_id",
+            existing_type=sa.VARCHAR(length=8),
+            type_=sa.Enum(
+                "Benevole", "LateUnregisterWarning", "Banned", name="badgeids"
+            ),
+            existing_nullable=False,
+        )
```
[the remark this answers is not in the document]
Comment (gdaviet) [reply migrations/versions/01e5792cdfac_add_registration_time.py:29]: Pardon, ce que je voulais dire c'est également mettre à jour le nom du fichier et le commentaire en haut du fichier pour être plus indicatif du contenu de la migration

**Comment by gdaviet** [comment]
Follow-up in #741

**Pull request closed without merging**